Some refactoring for pydantic models:

fselmo opened the pull request:
### What was wrong?

- Omit problematic fields from the json schema. This was causing
  some issues with fast api for some users that are serving the models
  over REST.

- Better organize and simplify the serialization of these models. Models
  should be indepdendently serializable and deserializable, as well
  as when creating a ``TypedTransaction.from_dict()``, NOT just when
  creating the ``TypedTransaction`` object. There was an issue where
  an ``@field_serializer`` was not being applied correctly because it was
  below the ``@classmethod``. It needs to be above the ``@classmethod``.
  This led to some unnecessary gymnastics to make the models work (see:
  ``recursive_model_dump()``). This method should probably be deprecated
  eventually.

- Add tests for serialization of the new datastructures / pydantic models
  and for generating the json schema.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-account/blob/main/newsfragments/README.md)

#### Cute Animal Picture

`0_o`

fselmo committed:
Some refactoring for pydantic models:

- Omit problematic fields from the json schema. This was causing
  some issues with fast api for some users that are serving the models
  over REST.

- Better organize and simplify the serialization of these models. Models
  should be indepdendently serializable and deserializable, as well
  as when creating a ``TypedTransaction.from_dict()``, NOT just when
  creating the ``TypedTransaction`` object. There was an issue where
  an ``@field_serializer`` was not being applied correctly because it was
  below the ``@classmethod``. It needs to be above the ``@classmethod``.
  This led to some unnecessary gymnastics to make the models work (see:
  ``recursive_model_dump()``). This method should probably be deprecated
  eventually.

- Add tests for serialization of the new datastructures / pydantic models
  and for generating the json schema.

fselmo commented on eth_account/datastructures.py:
```diff
@@ -153,13 +171,11 @@ class SignedSetCodeAuthorization(CustomPydanticModel):
+    authorization_hash: HexBytes = Field(exclude=True)
 
-    @classmethod
     @field_serializer("address")
```
keep this one in mind... this was the cause of unnecessary pain :)

royalnine reviewed (commented):
thank you @fselmo 🙏

reedsa commented on tests/core/test_datastructures.py:
```diff
@@ -0,0 +1,70 @@
+    (
+        (
+            TEST_SIGNED_AUTHORIZATION,
+            {
```
Nit: Expected data could be pulled into _test_utils

fselmo replied:
def worth it as it's already used in two places 👍🏼 ... updated

reedsa reviewed (approved):
No concerns, minor nit but up to you if it's worth using a constant.

fselmo committed:
changes from comments on PR #318:

- DRY: refactor expected JSON objects to test utils for
  use across multiple tests
